docs/v2_design.md

```diff
@@ -2,188 +2,600 @@
-```
 
----
+## Quantity Characteristics
 
-## Units
+ISO 80000 defines not only scalar quantities. Some of them are defined to have vector or tensor
+characteristics. To model that the library defines:
 
 ```cpp
-// Different magnitudes
-static_assert(si::metre != si::kilometre);
-static_assert(!equivalent(si::metre, si::kilometre));
-static_assert(convertible(si::metre, si::kilometre));
-
-// Unnamed derived unit vs named unit
-static_assert(si::newton * si::metre != si::joule);
-static_assert(equivalent(si::newton * si::metre, si::joule));
-static_assert(convertible(si::newton * si::metre, si::joule));
+enum class quantity_character { scalar, vector, tensor };
 ```
 
----
+The above assumes the following hierarchy:
+- every quantity can be expressed with a scalar representation type
+- vector representation type can be provided only to quantities with vector or tensor characteristics
+- tensor representation type can be used only for tensor quantities
 
-## Conversions
+The support for the above is still WIP. However, we assume that there will be variable template
+customization points for representation types that a user will have to specialize for a custom types.
 
-```cpp
-// Implicit conversions allowed between quantities of `convertible` references
-auto q1 = 120 * si::length[km] / (2 * si::time[h]);
-quantity<si::speed[km / h]> s = q1;
+The default values for type traits will be as follows:
 
-// Explicit casts allow changing all or only a part of the type
-auto q2 = quantity_cast<isq::speed_dim>(120 * si::length[km] / (2 * si::time[h]));
-auto q3 = quantity_cast<m / s>(120 * si::length[km] / (2 * si::time[h]));
-auto q4 = quantity_cast<si::speed[m / s]>(120 * si::length[km] / (2 * si::time[h]));
-auto q5 = quantity_cast<double>(120 * si::length[km] / (2 * si::time[h]));
-auto q6 = quantity_cast<quantity<si::speed[m / s], double>>(120 * si::length[km] / (2 * si::time[h]));
+```cpp
+template<typename T>
+inline constexpr bool is_scalar = false;
+
+template<typename T>
+  requires std::is_floating_point_v<T> ||
+             (std::is_integral_v<T> && one_of_types<T,
+                                                    signed char, short, int, long, long long,
+                                                    unsigned char, unsigned short, unsigned,
+                                                    unsigned long, unsigned long long>())
+inline constexpr bool is_scalar = true;
+
+template<typename T>
+inline constexpr bool is_vector = is_scalar<T>;
+
+template<typename T>
+inline constexpr bool is_tensor = is_vector<T>;
 ```
+
+Quantities will preserve their characteristics:
+- quantity kinds will have the same characteristics as the parent quantity (unless overridden)
+- derived quantities will have the most restrictive characteristics from its ingredients (unless overridden)
+
+It is still under analysis if the vector multiplied by the vector quantity should result
+in a tensor quantity or a vector one by default.
```

Comment:
Yeah, `*` can be overloaded. IIRC, one of the linear algebra proposals uses named functions and leaves the meaning of `*` to wrapper types. This is a non-issue with formulas. Perhaps `*` outside scalar arguments should require the specification of the characteristic.

Reply:
There are always exceptions and this is why we have overrides. Even your code explicitly put scalars in some quantities that are formed from vectors. Because this is their nature.

Be default I would like to assume that:
- `scalar * scalar` -> `scalar`
- `scalar * vector` -> `vector`
- `scalar * tensor` -> `tensor`
- `vector * tensor` -> `tensor`
- `tensor * tensor` -> `tensor`

The only one remaining that I am not sure about is `vector * vector`. Should be `vector` or `tensor` by default?

Reply:
There's also dot product, used in many quantity definitions, so it could also result in a scalar.

Reply:
At least until ISO 80000 part 10, tensors are rare, and I think I don't need 3 fingers to count the `vector * vector` -> `tensor` cases. The other two are much more common.